apps/remote_control/lib/lexical/remote_control/commands/rename.ex

```diff
@@ -0,0 +1,117 @@
+defmodule Lexical.RemoteControl.Commands.Rename do
+  @moduledoc """
+  We are unable to accurately determine the process of renaming,
+  because after the rename, there will be a series of operations such as
+  `DidChange`, `DidSave`, etc., which will trigger expensive actions like compiling the entire project.
+  Therefore, we need this module to make some markings to determine whether it is currently in the process of renaming.
+  """
+  defmodule State do
+    defstruct uri_with_operation_counts: %{}
+
+    def new(uri_with_operation_counts) do
+      %__MODULE__{uri_with_operation_counts: uri_with_operation_counts}
+    end
+
+    def mark_changed(%__MODULE__{} = state, uri) do
+      new_uri_with_operation_counts =
+        delete_key_or_reduce_counts(state.uri_with_operation_counts, uri)
+
+      %__MODULE__{state | uri_with_operation_counts: new_uri_with_operation_counts}
+    end
+
+    def mark_saved(%__MODULE__{} = state, uri) do
+      new_uri_with_operation_counts =
+        delete_key_or_reduce_counts(state.uri_with_operation_counts, uri)
+
+      %__MODULE__{state | uri_with_operation_counts: new_uri_with_operation_counts}
+    end
+
+    def mark_closed(%__MODULE__{} = state, uri) do
+      new_uri_with_operation_counts =
+        delete_key_or_reduce_counts(state.uri_with_operation_counts, uri)
+
+      %__MODULE__{state | uri_with_operation_counts: new_uri_with_operation_counts}
+    end
+
+    def in_progress?(%__MODULE__{} = state) do
+      state.uri_with_operation_counts != %{}
+    end
+
+    defp delete_key_or_reduce_counts(uri_with_operation_counts, uri) do
+      {_, new_map} =
+        Map.get_and_update(uri_with_operation_counts, uri, fn
+          nil -> :pop
+          current_counts when current_counts <= 1 -> :pop
+          current_counts -> {current_counts, current_counts - 1}
+        end)
+
+      new_map
+    end
+  end
+
+  use GenServer
+
+  def start_link(_) do
+    GenServer.start_link(__MODULE__, %State{}, name: __MODULE__)
+  end
+
+  @impl true
+  def init(state) do
+    {:ok, state}
+  end
+
+  def set_rename_progress(uri_with_operation_counts) do
```

Comment:
Note that when a large number of files are renamed, VSCode will delay sending some of the `DidClose` for old files and `DidOpen` for new files (for example, when renaming `Ast` -> `AST`, the last batch of messages will be received after 3-4 minutes). So, should we set a countdown here, for instance, reset the `uri_with_operation_counts` to `%{}` after 1 minute?

any ideas? @scohen